libexec/rpc.rquotad/rquotad.c

```diff
@@ -172,51 +205,81 @@ sendquota(struct svc_req *request, SVCXPRT *transp)
 }
 
 static void
-initfs(void)
+sendquota_extended(struct svc_req *request, SVCXPRT *transp)
 {
-	struct fstab *fs;
-
-	setfsent();
-	szqf = 8;
-	if ((qfa = malloc(szqf * sizeof *qfa)) == NULL)
-		goto enomem;
-	while ((fs = getfsent())) {
-		if (strcmp(fs->fs_vfstype, "ufs"))
-			continue;
-		if (nqf >= szqf) {
-			szqf *= 2;
-			if ((qfa = reallocf(qfa, szqf * sizeof *qfa)) == NULL)
-				goto enomem;
-		}
-		if ((qfa[nqf] = quota_open(fs, USRQUOTA, O_RDONLY)) == NULL) {
-			if (errno != EOPNOTSUPP)
-				goto fserr;
-			continue;
-		}
-		++nqf;
-		/* XXX */
+	struct ext_getquota_args getq_args;
+	struct getquota_rslt getq_rslt;
+	struct dqblk dqblk;
+	struct timeval timev;
+	int scale;
+
+	bzero(&getq_args, sizeof(getq_args));
+	if (!svc_getargs(transp, (xdrproc_t)xdr_ext_getquota_args, &getq_args)) {
+		svcerr_decode(transp);
+		return;
+	}
+	if (request->rq_cred.oa_flavor != AUTH_UNIX) {
+		/* bad auth */
+		getq_rslt.status = Q_EPERM;
+	} else if (!getfsquota(getq_args.gqa_type, getq_args.gqa_id, getq_args.gqa_pathp, &dqblk)) {
+		/* failed, return noquota */
+		getq_rslt.status = Q_NOQUOTA;
+	} else {
+		gettimeofday(&timev, NULL);
+		getq_rslt.status = Q_OK;
+		getq_rslt.getquota_rslt_u.gqr_rquota.rq_active = TRUE;
+		scale = 1 << flsll(dqblk.dqb_bhardlimit >> 32);
+		getq_rslt.getquota_rslt_u.gqr_rquota.rq_bsize =
+		    DEV_BSIZE * scale;
+		getq_rslt.getquota_rslt_u.gqr_rquota.rq_bhardlimit =
+		    dqblk.dqb_bhardlimit / scale;
+		getq_rslt.getquota_rslt_u.gqr_rquota.rq_bsoftlimit =
+		    dqblk.dqb_bsoftlimit / scale;
+		getq_rslt.getquota_rslt_u.gqr_rquota.rq_curblocks =
+		    dqblk.dqb_curblocks / scale;
+		getq_rslt.getquota_rslt_u.gqr_rquota.rq_fhardlimit =
+		    dqblk.dqb_ihardlimit;
+		getq_rslt.getquota_rslt_u.gqr_rquota.rq_fsoftlimit =
+		    dqblk.dqb_isoftlimit;
+		getq_rslt.getquota_rslt_u.gqr_rquota.rq_curfiles =
+		    dqblk.dqb_curinodes;
+		getq_rslt.getquota_rslt_u.gqr_rquota.rq_btimeleft =
+		    dqblk.dqb_btime - timev.tv_sec;
+		getq_rslt.getquota_rslt_u.gqr_rquota.rq_ftimeleft =
+		    dqblk.dqb_itime - timev.tv_sec;
+	}
+	if (!svc_sendreply(transp, (xdrproc_t)xdr_getquota_rslt, &getq_rslt))
+		svcerr_systemerr(transp);
+	if (!svc_freeargs(transp, (xdrproc_t)xdr_getquota_args, &getq_args)) {
+		syslog(LOG_ERR, "unable to free arguments");
+		exit(1);
 	}
-	endfsent();
-	return;
-enomem:
-	syslog(LOG_ERR, "out of memory");
-	exit(1);
-fserr:
-	syslog(LOG_ERR, "%s: %s", fs->fs_file, strerror(errno));
-	exit(1);
 }
 
 /*
  * gets the quotas for id, filesystem path.
  * Return 0 if fail, 1 otherwise
  */
 static int
-getfsquota(long id, char *path, struct dqblk *dqblk)
+getfsquota(int type, long id, char *path, struct dqblk *dqblk)
 {
-	int i;
+	struct quotafile *qf;
+	/*
+	 * This is STUPID, but otherwise when it
+	 * compiles with -Werrr,-Wcast-qual it fails.
+	 */
+	static char userquota[] = "userquota";
+	struct fstab fst = { .fs_file = path, .fs_mntops=userquota };
```

Comment:
This looks like a hack.  I suppose it may work for ZFS, which is never in fstab and always have quotas enabled, but for UFS I guess it may at least change behavior if real fstab does not have any quota options, while here you are hardcoding one.

Reply:
The particular layout is a hack, because of compiler options not allowing it to be written a sane way.  However, other than that, setting up the fst structure there is the right way:  this is rquotad, and it's being asked about a filesystem.  We can assume that it's mounted on the remote end, and that means it has to be mounted on the server end as well.  The quota_open() call will fail if quotas are not enabled on the mount point, so it all works out okay.

Reply:
I've gone through it again.  It does result in different behaviour, but I am going to maintain that the old behaviour -- which only looks at fstab, not mnttab -- is wrong.  Now, that gets to a question of whether zfs should add a generated "userquotas" and/or "groupquotas" for the mnttops, and ... that's a pretty good question.  Then it would instead be a matter of getting the mount options for a path at runtime, and determining whether it's set there.

Reply:
This needs review/comment still.  No response to my comment above, which was an implied question.

Options are to stall this until I make zfs mounting set userquotas and groupquotas, or accept it into freenas but not push it to freebsd, or to accept the change in behaviour and try to push it to freebsd.